### DISCUSSION TITLE
results about fine-tuning ESM2 model

### DISCUSSION THREAD

**Discussion by xtzhang0216** [facebookresearch/esm#556] [Q&A]
First of all congratulations on this fantastic work and publication.

I attempted to fine-tune the parameters of the last layer of the ESM2 model(3B parameters) on a specific dataset using huggingface. But during fine-tuning process, the loss only decreased from 1.25 to 1.2 (corresponding to a perplexity decrease from 3.5 to 3.32). According to Table S1 in supplementary materials (https://www.science.org/doi/10.1126/science.ade2574), I think this is a strange phenomenon because the perplexity(3.5)  during the first step of fine-tuning was much lower than the perplexity reported in the literature(6.49). Moreover, the fine-tuned model performs worse in the task of contact prediction as well.

I am highly suspicious that there might be issues with my fine-tuning code.
```python
from transformers import AutoTokenizer, DataCollatorForLanguageModeling
tokenizer = AutoTokenizer.from_pretrained("facebook/esm2_t36_3B_UR50D")
model = AutoModelForMaskedLM.from_pretrained("facebook/esm2_t36_3B_UR50D")
data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm_probability=0.15)
train_tokenized = tokenizer(train_sequences)
train_dataset = Dataset.from_dict(train_tokenized)
train_args = TrainingArguments(
    output_dir=args.save_folder,
    evaluation_strategy = "epoch",
    save_strategy = "epoch",
    learning_rate=5e-4,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=20,
    weight_decay=0.01,
    load_best_model_at_end=True,
)
trainer = Trainer(
    model,
    train_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
    tokenizer=tokenizer,
    data_collator=data_collator,
)
trainer.train()
```

 I am not sure if the low perplexity is due to the training batch size being selected based on the number of proteins instead of the number of tokens, or some other issue.
Hope for your suggestions. Thanks in advance!

**Comment by CrisSherban** [comment]
Hi @xtzhang0216, 
I am not an expert but I have also played with fine-tuning the HF model. 

I have a couple of notes with regard to this, let me know if you find them useful.

How big is the dataset on which you are fine-tuning?  
ESM2 3B might be an overkill to try fine-tuning if you have a small dataset, have you tried to fine-tune the smaller models first? 

Bach size can usually be very big in BERT, if I am not mistaken in ESM 2 there are 2 million tokens as batch size except the 15B. I guess this would be slightly disruptive for the model's loss to have a huge change in batch size.

Did you remember to manually freeze some layers with something like this:
```python
for param in model.base_model.encoder.parameters():
    param.requires_grad = False
```

Did you calculate perplexity on the same dataset as in the paper?


I hope this can help to brainstorm.

**Reply by xtzhang0216** [reply]
Thanks for your reply!
I am fine-tuning the ESM2 3B model in a database of approximately 30,000 specific protein sequences, with 90% used as training data and the remaining 10% as the test set.
```python
from sklearn.model_selection import train_test_split
from datasets import Dataset
train_sequences, test_sequences= train_test_split(seq, test_size=0.1, shuffle=True)
tokenizer = AutoTokenizer.from_pretrained("facebook/esm2_t36_3B_UR50D")
train_tokenized = tokenizer(train_sequences)
test_tokenized = tokenizer(test_sequences)
train_dataset = Dataset.from_dict(train_tokenized)
test_dataset = Dataset.from_dict(test_tokenized)
```
This is indeed a relatively small dataset, which leads to overfitting very quickly during finetuning. As I need to use the 3B model, I have not tried debugging on a small model yet and I am also in the process of collecting a larger dataset.

 What confuses me is that my perplexity is about half of the reported literature(3.5 vs 6.49) during the first firststep of fine-tuning the model. In other words, when I use ESM2 original parameters to evaluate in my specific dataset, the perplexity is too low.  As a beginner, I am very afraid that it may be my incorrect use of huggingface for fine-tuning (such as calculating loss incorrectly). 

As for the 2M tokens you mentioned, I'm not sure if I understand it correctly. ESM2 wants to concatenate sequences of different proteins into a single sequence of 2 million amino acids in total length. The different proteins are separated by 'bos''eos'. At this time, the batch size can be set to 1. Am I understanding this correctly?
But cross-entropy loss is calculated for each amino acid, whether it is based on 2M tokens or the 16 protein batches that I use, the loss is averaged per amino acid. The losses they calculated should be similar, unless the two tasks have different difficulties, which leads to different perplexities(may this be the true reason?) 

I have manually frozen some layers with something like that. I calculated perplexity in my test set, not the dataset in the paper.

Looking forward to your reply! Thanks in advance!

**Reply by avilella** [reply]
Hi @xtzhang0216 , I am interested in similar approaches to the one you describe here. Is there a way to direct message you? Thx in advance.

**Reply by xtzhang0216** [reply]
No, I'm just guessing based on training with 2 million tokens you mentioned. I don't know if my guess is right or not.
